### DISCUSSION TITLE
rotary dimmer rules sintaxis

### DISCUSSION THREAD

**Discussion by ximonline** [arendst/Tasmota#10511] [Q&A] [answered]
Hello! Could somebody help me with rules:

I use H801 for dimming LED stripes in WWWWW mode. For rotary encoder I use this rule:
rule1 on rotary1#pos1 do channel3 + endon
And it's particulary works (channel3 led stripe encrease for 10 by rotating encoder any side)

What rule I need for: turning right encrease led level, turning left decrease led level?

I think I need value of "rotary1#pos1" and "rotary1#pos2", but can't write it...

How can I see pos1 and pos2 values in console?

**Comment by barbudor** [comment, marked as answer]
You can't use + and - with `Rotary1#Pos1`
Pos1 (or Pos2) provide an absolute position between `0` and `ROTARY_MAX_STEPS`.
So you should use `on rotary1#pos1 do color %value%`

See example in https://tasmota.github.io/docs/Rules/#use-a-rotary-encoder

**Reply by barbudor** [reply]
If you want to see values in console, the simplest is to make a rule that assign the value to a var
`on rotary1#pos1 do var1 %value% endon on rotary1#pos2 do var2 %value% endon`

**Comment by Jason2866** [comment]
This should work: 
`SetOption98 1`
`rule1 on rotary1#Pos1 do dimmer - endon on Rotary1#Pos2 do do dimmer + endon`

**Reply by barbudor** [reply]
When using SO98 (which was assumed in my response above as it is stated in the linked documentation)
- Pos1 is managed by default
- Pos2 is managed is Switch1 is pressed

The idea is to use for example
- turn no-pressed for luminance
- turn pressed for color 

Both Pos1 and Pos2 are absolute and will inc/dec depending on the rotation direction. They will be capped with the the 0 .. ROTARY_MAX_STEPS range
There is no modes to get inc/dec information, only absolute. Would be nice though.

**Reply by ximonline** [reply]
> There is no modes to get inc/dec information, only absolute. Would be nice though.

Should I write somethin about:

Var1=pos1
On rotary1#pos1 do (if  pos1>var1 do dimmer +  else do dimmer - ) var1= pos1 endon

How can I get this in tasmota rules?..

**Reply by barbudor** [reply]
No rule could work because when the `Pos1` is at 0, turning it more on the left won't go to negative.
Same on the other direction with ROTARY_MAX_STEPS

**Reply by barbudor** [reply]
You could write rules like :
```
on rotary1#pos1>%var1% do backlog var1 %value% ; dimmer + endon
on rotary1#pos1<%var1% do backlog var1 %value% ; dimmer - endon
```
And recompile with a huge ROTARY_MAX_STEPS such as 16000 or 32000
But at some point you may hit the walls

**Comment by arendst** [comment]
> And recompile with a huge ROTARY_MAX_STEPS such as 16000 or 32000

Won't work with latest development as rotary rules now use ``SetOption43 1..100``. Here 1..100 is the "old" ROTARY_MAX_STEPS value.

**Reply by barbudor** [reply]
So ROTARY_MAX_STEPS  define is dropped in favor of SO43, up to 255 (default 100) ?

The good thing with Tasmota is that whatever you consider as given won't be true next week.
Continuous learning 🤣

**Reply by ximonline** [reply]
> 
> Won't work with latest development as rotary rules now use `SetOption43 1..100`. Here 1..100 is the "old" ROTARY_MAX_STEPS value.

I use tasmota 9.2.0 release, setoption43 is "100".
But rule "ON Rotary1#Pos1 DO channel4 %value% ENDON" works only in 1.. 10 range...

**Reply by barbudor** [reply]
As it has been changed on the last days, you need the development version 9.2.0.3.
You can upgrade seamlessly with Ota using URL `http:/ota.tasmota.com/tasmota/tasmota.bin.gz`

**Reply by barbudor** [reply]
But as pointed by Theo, max is 255

**Comment by arendst** [comment]
> So ROTARY_MAX_STEPS define is dropped in favor of SO43, up to 255 (default 100) ?

It's a bit quirky (as most changes I make these days) but using SO43 you can only set max_steps between 1 and 100 OR (tada) to ROTARY_MAX_STEPS which is supported up to 255 (as max_steps is a 8-bit register).

Now I think about it this change was implemented to satisfy both Mi-Led and Generic module support. Using rules it makes more sense to be able to set it to the full range of 1 to 255 using SO43....

I'll change it now to allow setting it to 255 with SO43.

**Reply by arendst** [reply]
Latest development now supports ``SO43 1..255``

**Reply by barbudor** [reply]
understood. Thanks Theo.
(although encoders with more than 100 steps per rotation probably not common)

**Comment by ximonline** [comment]
With 9.2.0.3 release dimmer works good!
Additional question about it:
 Default value of button1 push maped at channel1 switch. How can I remap button1?

**Reply by barbudor** [reply]
What do you want to do with button1 ?
You can generally override button behavior with rules

**Reply by ximonline** [reply]
> What do you want to do with button1 ?
> You can generally override button behavior with rules

Thnx for you unswers. Could you tell me write sintax for overriding button1 for another channel switch off/on

**Reply by barbudor** [reply]
Read the Rules chapter at https://tasmota.github.io/docs/Rules/

A rule to override a button action will mostly looks like `on button1#state do .... endon`

Please read the docs

**Comment by brettlg** [comment]
Hi All,
I must be missing something obvious.  I can get rotary encoder working with dimmer but I cannot for the life of me get it to dim to 0 (off) by turning the knob.  I want to use the push button on my encoder for another purpose so wanting to dim to off with the rotary encoder.  Minimum I can get is 1.  Also, just a little thing but the change to dim level is not showing in the GUI without refresh, it does however show in console. Ideally I would like dimmer range to be 0 or 10-100% as the LEDs don't perform well under 10%.  My rules are as follows...

Rule2 on Rotary2#Pos1==1 do dimmer + endon on Rotary2#Pos1==0 do dimmer - endon 

Any help much appreciated
Thanks

**Reply by sfromis** [reply]
You tell very little of what you configured, and why you even have that rule set for rotary dimmer operation. Not needed for the built-in feature.

Values of SetOption43, SetOption77, SetOption98, SetOption113 make a major diffrence to operation.

**Reply by brettlg** [reply]
Thanks for your lightning fast reply.  

I have a Moes MS-105-1 that I have replaced the existing non flashable chip with a ESP-02S and broken out unused gpio's.
Have been using a separate device running ESPurna to handle the encoder (this was setup before TASMOTA had such good encoder support) with the upgraded Moes working as expected but am wanting to decrease the device count and potentially improve responsiveness with commands handled  locally on the one device (also space saving) .  

**Using following configuration....** 
{"NAME":"MS-105","GPIO":[0,107,0,108,0,0,0,0,0,0,0,0,0],"FLAG":0,"BASE":54}
TuyaMCU 21,2
DimmerRange 0,720
SetOption113 off, SetOption43 1, SetOption98 on
Rule2 on Rotary2#Pos1==1 do backlog dimmer +5 endon on Rotary2#Pos1==0 do backlog dimmer - endon 

Encoder is working but only ranges 1 - 100.  Ideally if less that 10% turn off otherwise dim as per direction of encoder.

I'm using a KY-040 encoder, push button I want to use to control fan speed on an iFan02 also running TASMOTA.  Each button press decreasing fan speed by one so 0, 3, 2, 1, 0 (wanting to keep knob/button count to a minimum and keep operation as guest/senior citizen proof as possible.

I feel somewhat stupid that I managed to get 99% of the way there but am humbled by the help available on the community.

Thanks again for your help.  I think that's all the settings, let me know if there is anything else you need.

**Reply by sfromis** [reply]
You did not answer what you have in `SetOption77`, which could key to your issue with not dimming down to off.

Anyway, Tasmota does not have full control when you are using a TuyaMCU-infested device, as Tasmota basically only has what a smartphone app could do.

**Reply by brettlg** [reply]
TuyaMCU-infested device LOL

RESULT = {"SetOption77":"OFF"} tried with on and it would not power down in TASMOTA GUI so returned it to off.  I got excited when I saw that option, overlooked it from your previous message sorry.

Light will power off  when when using same encoder on ESPurna and also in tasmota GUI.

For some strange reason the encoder (when driven by TASMOTA) will not supply a value less than 1 

My rules ability is somewhat NOOB.
Would it be plausible that a rule could state if dimmer.value less than 10 then dimmer  0 else add subtract or am I still missing something simple and overthinking it?  

I am so close to TASMOTA Dimmer heaven

Prototype one, version 2 should have no solder iron scorch marks on case

TX, RX, GPIO0 and 5 other GPIOs and 3.3v on exterior headers

![IMG_5278](https://github.com/arendst/Tasmota/assets/30974297/3cd3d831-8694-4423-b55a-0ca4b432066e)
![IMG_5274](https://github.com/arendst/Tasmota/assets/30974297/774a39a7-1b8a-4f0d-ac10-503f3aaca476)

**Reply by brettlg** [reply]
You got me thinking from some of your previous comments and by some fluke I ended up setting rotary a/b 1 (vs 2 previously to potentially capture a button press with rotary) and turned off rules and setoption43 100.  Now is dimming from 0 - 100 but takes multiple turns of the encoder as it counts +- 1.   Can this be speed up to count in 5 or 10 or is this a limitation of  TuyaMCU-infested device?

I have tried changing a few things but I'm shooting in the dark here ie SetOption43 50, DimmerStep 10, have also changed Base setting but don't really know what I'm doing there

Again your help is much appreciated,  you managed to get the wheels turning in a better direction.

**Reply by sfromis** [reply]
Well, if you want less twisting, your rather high values for SetOption43 may be counter-productive. The default is 16, but it might work better with a value around 10. The rotary encoder driver can go down to 0, no problem, this is what I see with a "native" dimmer on a device with a rotary encoder, but this is not including the TuyaMCU games.

One possible way would be to set up a rule to issue a `Dimmer 0` when you are in a situation where you'd want to turn off, but this depends a lot of what signals you have in the JSON payloads to recognize the situation. The command `Dimmer -` will not go below 1. If nothing better, you could go back to issuing Dimmer commands as reaction to the rotary, and `Dimmer 0` when relevant.

**Reply by brettlg** [reply]
Cheers.  That's what I found with SetOption43 1 & SetOption98 on using rule, only values of 1 - 100 but I could configure steps of 10 or whatever by changing dimmer + value or by changing DimmerStep to whatever I wanted if using rules.  I tried changing SetOption43 to value of 10 with SetOption98 off and rules off but this limited the dimmer range to a maximum percentage of what I put as the value.

I feel now I'm within spiting distance of the Tasmota-Dimmer heaven and will keep tweaking

**Reply by brettlg** [reply]
Success.  I have entered TASMOTA - TuyaMCU-infested Dimmer Nirvana.

Have finally won with ....
{"NAME":"MS-105","GPIO":[0,107,0,108,0,0,0,0,0,0,0,0,0],"FLAG":0,"BASE":54}
TuyaMCU 21,2
DimmerRange 0,720
SetOption113 off, SetOption43 1, SetOption98 on, DimmerStep 10 (or whatever other valuse), SetOption69 1

Rule1 on Rotary2#Pos1==1 do dimmer + endon 
          on Rotary2#Pos1==0 do dimmer - endon 
          on Dimmer#State==1 do dimmer 0 endon 
2 last questions, can debounce value  for the encoder be changed or is this set at compile time and do you know where there  is there more info on what the different Base selections change Base 54, Base 18 etc?

Thanks again for putting up with me

**Reply by sfromis** [reply]
https://tasmota.github.io/docs/Modules/

And `ROTARY_DEBOUNCE` can only be set build time.
